c++/src/kj/compat/http.c++

```diff
@@ -2295,12 +2298,25 @@ class WebSocketImpl final: public WebSocket {
 public:
   WebSocketImpl(kj::Own<kj::AsyncIoStream> stream,
                 kj::Maybe<EntropySource&> maskKeyGenerator,
+                kj::Maybe<CompressionParameters> compressionConfig = nullptr,
                 kj::Array<byte> buffer = kj::heapArray<byte>(4096),
                 kj::ArrayPtr<byte> leftover = nullptr,
                 kj::Maybe<kj::Promise<void>> waitBeforeSend = nullptr)
       : stream(kj::mv(stream)), maskKeyGenerator(maskKeyGenerator),
+        compressionConfig(kj::mv(compressionConfig)),
         sendingPong(kj::mv(waitBeforeSend)),
-        recvBuffer(kj::mv(buffer)), recvData(leftover) {}
+        recvBuffer(kj::mv(buffer)), recvData(leftover) {
+    #if KJ_HAS_ZLIB
+    KJ_IF_MAYBE(config, this->compressionConfig) {
+      compressionContext.emplace(ZlibContext::Mode::COMPRESS, *config);
+      decompressionContext.emplace(ZlibContext::Mode::DECOMPRESS, *config);
+    }
+    #else
+    this->compressionConfig = nullptr;
```

Comment:
I think we should be KJ_REQUIREing that this is null, rather than changing it to null? If it is non-null then we've already negotiated compression settings we can't support.

On another note, usually if I find myself using `this->blah` in a constructor to avoid confusion with a parameter, I instead rename the parameter to `blahParam` so that `blah` unambiguously refers to the member.

Reply:
I might be totally off here and/or overthinking this, but suppose I wanted to use KJ websockets for proxying and I didn't build with `zlib`. Let's also say I still want the client/server to use compression. In this case, couldn't the client/server technically use compression while the proxy just does an optimized pump? In that case, doesn't it sort of make sense to actually allow non-null `compressionConfig`s?

I guess what I'm asking is would it make more sense to:
- Not force `compressionConfig` to be null here
- `KJ_REQUIRE` the config to be null if we try to _unoptimized_ pump using websocket compression (when zlib hasn't been linked)

Also noted for the param, I'll fix that. Thanks for mentioning because IIRC I got confused by why I had to use `this`.

Reply:
Param name fixes:
- https://github.com/capnproto/capnproto/pull/1505/commits/58d08a749eb2d5d5db75dc304d99729d4f89562c
- https://github.com/capnproto/capnproto/pull/1505/commits/427686e6e01636dc3171b7d61fe508460c545232

Reply:
> Let's also say I still want the client/server to use compression. In this case, couldn't the client/server technically use compression while the proxy just does an optimized pump? In that case, doesn't it sort of make sense to actually allow non-null compressionConfigs?

In that case you would need to make sure the code throws an exception if you try to do anything other than an optimized pump. This does not seem worth the effort to me. I think it's fine if we say you can't have compression at all unless you compile with zlib. Compiling with zlib is not a high barrier anyway.

Reply:
https://github.com/capnproto/capnproto/pull/1505/commits/6dc671574cb7516a353781c17fafa11556d72932